### PR TITLE
vboxlauncher@mockturtl: add headless mode for VirtualBox machines

### PR DESCRIPTION
**Pull request by edaubert**
This pull request change the display to allow to start Virtualbox VMs using the headless mode

@mockturtl

### CODE REVIEW

File: vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js [hunk 1]
```diff
@@ -13,7 +13,7 @@ const ICON = "virtualbox"
 const CMD_VBOX = "virtualbox"
 const CMD_VBOXMANAGE = "vboxmanage"
 
-const CMD_VBOX_VM = CMD_VBOX + " --startvm "
+const CMD_VBOX_VM = CMD_VBOXMANAGE + " startvm "
```
Comment (mockturtl) [inline vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js]: This change makes `CMD_VBOX_VM` identical to `CMD_VBOX6_VM` (below).  Doesn't that break Virtualbox 5.x?
Notice vbox 6 support was introduced [here](https://github.com/linuxmint/cinnamon-spices-applets/commit/5751e8dd6fc92eb86c41da51b60f549d1117aff4#diff-41b8487c7ae21a4c12ec229ce7570857R17).

File: vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js [hunk 2]
```diff
@@ -52,7 +52,7 @@ MyApplet.prototype = {
 
     try {
       this.set_applet_icon_name(ICON)
-      
+
```
Comment (mockturtl) [inline vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js]: Ideally whitespace fixes should be in a separate commit, but it's OK.  

If you want to level up your git, you can practice:

```sh
$ git reset HEAD~  # bring the working directory to the state before this commit
$ git add --patch applet.js  # select changes that go together, instead of the whole file
                             # then commit as usual
```

File: vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js [hunk 3]
```diff
@@ -183,7 +191,7 @@ MyApplet.prototype = {
       name = (name+"   \uE226");
     }
     let id = info[1].replace('}', '')
-    this.addLauncher(name, Lang.bind(this, function() { this.startVboxImage(id) }))
+    this.addLauncher(name, Lang.bind(this, function() { this.startVboxImage(id) }), Lang.bind(this, function() { this.startVboxImageHeadless(id) }))
```
Comment (mockturtl) [inline vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js]: Need a line break here.

File: vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js [hunk 4]
```diff
@@ -133,11 +141,11 @@ MyApplet.prototype = {
       this.addErrorMessage(_("ERROR. No compatible virtual machine programs found."))
   }
 
-, launcherFor: function(title, cmd, callback) {
+, launcherFor: function(title, cmd, callback, callbackHeadless) {
```
Comment (mockturtl) [inline vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js]: Is there a typo here?  This function is only called in L135, L136, and neither callsite passes the new argument.

File: vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js [hunk 5]
```diff
@@ -237,12 +245,17 @@ MyApplet.prototype = {
...
-  
+
 ,  startVboxImage: function(id) {
     let cmd = this.vboxMajorVersion() >= 6 ? CMD_VBOX6_VM : CMD_VBOX_VM
     Util.spawnCommandLine(cmd + id)
   }
-  
+
+,  startVboxImageHeadless: function(id) {
+    let cmd = this.vboxMajorVersion() >= 6 ? CMD_VBOX6_VM : CMD_VBOX_VM
```
Comment (mockturtl) [inline vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js]: As noted above, the git history indicates vboxmanage <= 5 takes `--startvm` versus `startvm`.

Can we confirm it vbox <= 5 does work with `--type headless`?

File: vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js (hunk 4 above)
Comment (mockturtl) [reply vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js]: Now that I look at it, everything named "launcher" is really a menuItem.  We should change names to reflect that.

File: vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js (hunk 1 above)
Comment (edaubert) [reply vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js]: I have VitrualBox 5.2.20 and everything is fine.

File: vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js (hunk 5 above)
Comment (edaubert) [reply vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js]: Maybe someone else can confirm but acording to this tutorial, there is vboxmanage at least from 5.2
https://websiteforstudents.com/installing-virtualbox-5-2-ubuntu-17-04-17-10/

File: vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js (hunk 3 above)
Comment (edaubert) [reply vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js]: Done

File: vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js (hunk 2 above)
Comment (edaubert) [reply vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js]: You completely right, sorry, I rollback this kind of changes !

File: vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js (hunk 4 above)
Comment (edaubert) [reply vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js]: I remove the callbackHeadless as this method is only called to display the menuItem for Vbox and VMware.

Those launchers open the corresponding apps.
How would you like them to be called ?

File: vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js (hunk 1 above)
Comment (mockturtl) [reply vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js]: OK.

File: vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js (hunk 4 above)
Comment (mockturtl) [reply vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js]: It's fine for now.

File: vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js (hunk 5 above)
Comment (mockturtl) [reply vboxlauncher@mockturtl/files/vboxlauncher@mockturtl/applet.js]: Okay, I don't quite follow the sea of commands and symlinks in 5.x, but I see `--type headless` is supported.

```sh
 $ vboxmanage --help
Oracle VM VirtualBox Command Line Management Interface Version 5.2.24_Debian
(C) 2005-2019 Oracle Corporation
All rights reserved.
 ...
  startvm                   <uuid|vmname>...
                            [--type gui|sdl|headless|separate]
                            [-E|--putenv <NAME>[=<VALUE>]]
...
```

For reference, v6.0:

```sh
$ virtualbox --help
Oracle VM VirtualBox VM Selector v6.0.4
(C) 2005-2019 Oracle Corporation
All rights reserved.

No special options.

If you are looking for --startvm and related options, you need to use VirtualBoxVM.
$ vboxmanage startvm
Usage:

VBoxManage startvm          <uuid|vmname>...
                            [--type gui|sdl|headless|separate]
                            [-E|--putenv <NAME>[=<VALUE>]]
```

**Comment by mockturtl** [comment]
@jaszhix :+1:
